---
title: "Chronogram annotation"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Chronogram annotation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\VignetteDepends{stringr}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 6,
  fig.height = 8
)
```


## Introduction

The chronogram package provides a family of functions to annotate a chronogram. These all start `cg_annotate_`. This vignette explains how to use these annotation functions. Before using this vignette, consult the `vignette("assembly")`.

## Setup

```{r setup}
library(chronogram)
library(dplyr)
library(ggplot2)
library(patchwork)
```

We will use the example pre-built chronogram, introduced in the `vignette("assembly")`, and add on some example infection data.

```{r}
data(built_smallstudy)
cg <- built_smallstudy$chronogram

## Simulate some infection data ##
infections_to_add <- tibble::tribble(
  ~calendar_date, ~elig_study_id, ~LFT, ~PCR, ~symptoms,
  "01102020", "1", "pos", "not tested", NA,
  "11102020", "1", "pos", NA, "severe",
  "11042021", "2", "pos", "not tested", "mild",
  "11082021", "2", "pos", "Pos", "mild",
  "11112021", "2", "pos", "Pos", NA
)
## Make calendar_date a date ##
infections_to_add$calendar_date <- lubridate::dmy(
  infections_to_add$calendar_date
)
## add to chronogram
cg <- cg_add_experiment(
  cg,
  infections_to_add
)
```

## Outline

Annotation is required to allow the selection sub-cohorts of individuals (and corresponding dates) that are relevant to test your biological hypothesis.

### The annotation process

#### 1. For vaccines, use `cg_annotate_vaccines_count()`

-- adds a column which counts the number of vaccines each participant has received over time.

-- includes a "star" system, to allow the first few days after a vaccine to be easily identified. For example, 24hrs after the first dose of a vaccine ("1star") is probably not biologically reflect of that dose's effect. The user can set the value of days to "star" to suit their analysis.

#### 2. For episodes, use the `cg_annotate_episodes_` family

These perform sequential steps:

i. `cg_annotate_episodes_find()` is used to identify episodes of infection. This _reads_ across several different experiment data types, that are specified in the function call: eg. symptoms, PCRs, LFTs for SARS-CoV-2. Additionally, `cg_annotate_episodes_find()` looks backwards and forwards in time to find episodes, as commonly the molecular testing and symptoms do not commence on the same day.

In some studies, seroconversion is the method of finding episodes (eg seroconversion on anti-N IgG in SARS-CoV-2). Seroconversion episodes can be found with `cg_annotate_episodes_find_seroconversion()`. As these provide different (often much, much larger) uncertainty around the episode start, chronogram handles this in a separate function.

ii. `cg_annotate_episodes_fill()` allow filling of specific columns for each episode (details below).

iii. count the number of episodes (cf `cg_annotate_vaccines_count()`)

#### 3. For exposures, use `cg_annotate_exposure_count()` and `cg_annotate_antigenic_history()`

Exposures can be either from vaccine doses or from infection episodes. This function provides a cumulative counter for each individual's personal history.

## 1. Annotating vaccines

`cg_annotate_vaccines_count()` requires that metadata columns follow this pattern:

- dose_1, dose_2, dose_3, ..., dose_i

- date_dose_1, date_dose_2, date_dose_3, ..., date_dose_i

The {dose} and {date_dose} prefixes should be provided to the function, as shown in the chunk below. You might envisage a study with {sarscov2_dose}, {sarscov2_date_dose}, {influenza_dose} & {influenza_date_dose} for which two runs of `cg_annotate_vaccines_count()` would be needed.


```{r}
cg1 <- cg_annotate_vaccines_count(
  cg,
  ## the prefix to the dose columns: ##
  dose = dose,
  ## the output column name: ##
  dose_counter = dose_number,
  ## the prefix to the date columns: ##
  vaccine_date_stem = date_dose,
  ## use 14d to 'star' after a dose ##
  intermediate_days = 14
)

## plot over time ##
cg1 %>%
  ggplot(
    aes(
      x = calendar_date,
      y = elig_study_id,
      fill = dose_number
    )
  ) +
  geom_tile(height = 0.5) +
  scale_fill_grey(end = 0.2, start = 0.8) +
  theme_bw()
```

In the above plot, the character vector `dose_number` is coerced to a factor. The resulting levels of the factor are counter-intuitive. To improve the plot, you can manually specify the factor: 

```{r}
cg1 %>%
  mutate(dose_number = factor(dose_number,
    levels = c(
      "0",
      "1star",
      "1",
      "2star",
      "2"
    )
  )) %>%
  ggplot(
    aes(
      x = calendar_date,
      y = elig_study_id,
      fill = dose_number
    )
  ) +
  geom_tile(height = 0.5) +
  scale_fill_grey(end = 0.2, start = 0.8) +
  theme_bw()
```

The `dose_number` column is retained as character vector rather than converting to a factor. There is the possibility for mishandling if a factor: `as.numeric(dose_number) == 2`, refers to the situation `dose == 1`.

## 2. Annotating episodes

The first step is to find episodes using `cg_annotate_episodes_find()`

```{r}
cg2 <- cg_annotate_episodes_find(
  cg1,
  infection_cols = c("LFT", "PCR", "symptoms"),
  infection_present = c("pos", "Post", "^severe")
)

cg2 %>%
  mutate(dose_number = factor(dose_number,
    levels = c("0", "1star", "1", "2star", "2")
  )) %>%
  ggplot(
    aes(
      x = calendar_date,
      y = elig_study_id,
      fill = dose_number
    )
  ) +
  geom_tile(height = 0.5) +
  ## now add text to show when episodes occur ##
  geom_text(
    aes(label = episode_number),
    data = . %>%
      group_by(elig_study_id, episode_number) %>%
      slice_head(),
    col = "red", show.legend = FALSE
  ) +
  scale_fill_grey(end = 0.2, start = 0.8) +
  theme_bw()
```

Note that participant 1 has a single infection episode in November 2020:

```{r}
infections_to_add %>%
  filter(elig_study_id == 1) %>%
  knitr::kable()
```

We could use a shorter time frame to look forwards and backwards in:

```{r}
cg3 <- cg_annotate_episodes_find(
  cg1,
  infection_cols = c("LFT", "PCR", "symptoms"),
  infection_present = c("pos", "Post", "^severe"),
  ## change episode days interval to 7 days ##
  episode_days = 7
)

cg3 %>%
  mutate(dose_number = factor(dose_number,
    levels = c("0", "1star", "1", "2star", "2")
  )) %>%
  ggplot(
    aes(
      x = calendar_date,
      y = elig_study_id,
      fill = dose_number
    )
  ) +
  geom_tile(height = 0.5) +
  ## now add text to show when episodes occur ##
  geom_text(
    aes(label = episode_number),
    data = . %>%
      group_by(elig_study_id, episode_number) %>%
      slice_head(),
    col = "red", show.legend = FALSE
  ) +
  scale_fill_grey(end = 0.2, start = 0.8) +
  theme_bw()
```

The episode_days option sets the "gap" to look for, rather than the total permitted length of an episode. For example, LFT+ve on d-6, symptoms on d0, and PCR+ on d7 would be collected into a single episode.

Next, we assign variants. There is no one-size-fits all approach that holds across pathogens, across studies, and across geographic regions. In this example we have PCR data available, and could parse as: if there is PCR positivity assign Delta, if there is no PCR testing assign Ancestral/Alpha. 

```{r}
cg4 <- cg2 %>%
  mutate(
    episode_variant =
      case_when(
        # "is an episode" & "PCR positive" -> Delta #
        (!is.na(episode_number)) & PCR == "Pos" ~ "Delta",
        # "is an episode" & "PCR unavailable" -> Anc/Delta #
        (!is.na(episode_number)) & PCR == "not tested" ~ "Anc/Alpha"
      )
  )
```

This gives a variant call on a single row of `cg`.

```{r}
cg4 %>%
  filter(!is.na(episode_variant)) %>%
  pull(episode_variant)
```

We can use `cg_annotate_episodes_fill()` to add this annotation to every row of the infection episode.

```{r}
cg4 <- cg4 %>%
  cg_annotate_episodes_fill(
    col_to_fill = episode_variant,
    col_to_return = episode_variant_filled,
    .direction = "updown",
    episode_numbers_col = episode_number
  )


cg4 %>%
  filter(!is.na(episode_variant_filled)) %>%
  pull(episode_variant_filled)
```

Variant calling rarely relies on a single assay. We advocate a pragmatic a tiered approach. If viral sequencing results are available we use those; if PCR only (Â± S gene target failure), we use those; and in the absence of molecular testing we compare episode start dates to prevailing variant trends. Give each their own `episode_variant_by_seq`, `episode_variant_by_PCR` and `episode_variant_by_date`, and then fill `*_filled` for the whole episode. And then apply a further `case_when()` to combine over `episode_variant_by_seq_filled`, `episode_variant_by_PCR_filled` and `episode_variant_by_date_filled`. We adopt `episode_variant_summarised` as the combined, overall assignment, but the user is free to change.

Sequencing interpretation is its own sub-discipline, which we have simplified in this vignette. For SARS-CoV-2, there are two main methods of assigning a variant: PANGO and scorpio, and these, on occasion, report different lineages particularly for newly emerging viruses. We keep both: `episode_variant_by_pango` and `episode_variant_by_scorpio`. 

Finally, we can count the number of episodes that each individual has experienced.

```{r}
cg4 %>%
  cg_annotate_episodes_count() %>%
  pull(count_col)
```


## 3. Annotating exposures

Annotation of exposures is the cumulative sum of infection + vaccination histories. These annotations are provided by `cg_annotate_exposure_count()` and `cg_annotate_antigenic_history()`.

`cg_annotate_exposure_count()` takes the column names for episode number, dose number and seroconversion episode numbers, to calculate a per day, per individual running count of exposures.

`cg_annotate_antigenic_history()` provides a character vector of length 1 to summarise each person's course over the study.

These are both best explored with an example, and a plot.

```{r}
cg_exposures <- cg4 %>% cg_annotate_exposures_count(
  episode_number = episode_number,
  dose_number = dose_number,
  ## we have not considered episodes of seroconversion
  N_seroconversion_episode_number = NULL
)


cg_exposures <- cg_exposures %>%
  mutate(
    episode_variant_summarised =
      episode_variant_filled
  ) %>%
  cg_annotate_antigenic_history(
    episode_number = episode_number,
    dose_number = dose_number,
    episode_variant_summarised = episode_variant_summarised,
    ag_col = antigenic_history
  )

## Plot ##
top_panel <- cg_exposures %>%
  select(calendar_date, 
         exposure_number,
         elig_study_id,
         antigenic_history) %>%
  ggplot(aes(
    x = calendar_date, y = exposure_number,
    col = elig_study_id
  )) +
  geom_line() +
  facet_grid(antigenic_history ~ .)


swimmers_panel <- cg_plot_meta(cg_exposures,
  visit = serum_Ab_S
) +
  ## set the axes to match top_panel ##
  xlim(
    min(cg_exposures$calendar_date),
    max(cg_exposures$calendar_date)
  ) +
  scale_y_discrete(limits = factor(c(3, 2, 1)))

top_panel / swimmers_panel & theme_bw() &
  theme(
    legend.position = "bottom",
    strip.text.y = element_text(angle = 0),
    strip.background = element_blank()
  )
```

The plot above shows how each infection, or vaccination impacts the exposure number. For participant 3, their only exposures are vaccination, whereas both participants 1 and 2 have additional exposures from infection. By the end of this example, participant 2 has experienced 4 encounters with Spike, participant 1 has had 3 encounters and participant 3 just two from vaccination alone.

# Summary

This vignette has provided examples of the cg_annotate family in action. If you are conducting a multi-pathogen study (RSV, flu, covid), then run a set of cg_annotate family functions for each pathogen - and you may wish to prefix the output columns eg `RSV_`,  `flu_` & `covid_`. As these have differing considerations for eg variants, chronogram leaves the cg_annotate family without an overall wrapper to let users easily omit unneeded annotations.

## SessionInfo

```{r}
sessionInfo()
```
