---
title: "Annotate episodes"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Annotate episodes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\VignetteDepends{stringr}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 6,
  fig.height = 8
)
```


## Introduction

The chronogram package provides a family of functions to annotate a chronogram. These all start `cg_annotate_`. This vignette explains how to use these annotation functions. Before using this vignette, consult the `vignette("assembly")`.

## Setup

```{r setup}
library(chronogram)
library(dplyr)
library(ggplot2)
library(patchwork)
```

We will use the example pre-built chronogram, introduced in the `vignette("assembly")`, and add on some example infection data.

```{r}
data(built_smallstudy)
cg <- built_smallstudy$chronogram
infections_to_add <- built_smallstudy$infections_to_add

## add to chronogram
cg <- cg_add_experiment(
  cg,
  infections_to_add
)
```

## Episode annotation

Annotation is required to allow the selection sub-cohorts of individuals (and corresponding dates) that are relevant to test your biological hypothesis.

For episodes, use the `cg_annotate_episodes_` family

These perform sequential steps:

i. `cg_annotate_episodes_find()` is used to identify episodes of infection. This _reads_ across several different experiment data types, that are specified in the function call: eg. symptoms, PCRs, LFTs for SARS-CoV-2. Additionally, `cg_annotate_episodes_find()` looks backwards and forwards in time to find episodes, as commonly the molecular testing and symptoms do not commence on the same day.

In some studies, seroconversion is the method of finding episodes (eg seroconversion on anti-N IgG in SARS-CoV-2). Seroconversion episodes can be found with `cg_annotate_episodes_find_seroconversion()`. As these provide different (often much, much larger) uncertainty around the episode start, chronogram handles this in a separate function.

ii. `cg_annotate_episodes_fill()` allow filling of specific columns for each episode (details below).

iii. count the number of episodes (cf `cg_annotate_vaccines_count()`)

### Worked example

The first step is to find episodes using `cg_annotate_episodes_find()`.

```{r}
cg_episodes <- cg_annotate_episodes_find(
  cg,
  infection_cols = c("LFT", "PCR", "symptoms"),
  infection_present = c("pos", "Post", "^severe")
)

cg_episodes %>%
  ggplot(
    aes(
      x = calendar_date,
      y = elig_study_id,
      label = episode_number
    )
  ) +
  ## now add text to show when episodes occur ##
  geom_text(
    data = . %>%
      group_by(elig_study_id, episode_number) %>%
      slice_head(),
    col = "red", show.legend = FALSE
  ) +
  theme_bw()
```

Note that participant 1 has a single infection episode in November 2020, with testing and symptom data on different days:

```{r}
infections_to_add %>%
  filter(elig_study_id == 1) %>%
  knitr::kable()
```

We could use a shorter time frame to look forwards and backwards in:

```{r}
cg_short_episode_length <- cg_annotate_episodes_find(
  cg,
  infection_cols = c("LFT", "PCR", "symptoms"),
  infection_present = c("pos", "Post", "^severe"),
  ## change episode days interval to 7 days ##
  episode_days = 7
)

cg_short_episode_length %>%
  ggplot(
    aes(
      x = calendar_date,
      y = elig_study_id,
      label = episode_number
    )
  ) +
  ## now add text to show when episodes occur ##
  geom_text(
    data = . %>%
      group_by(elig_study_id, episode_number) %>%
      slice_head(),
    col = "red", show.legend = FALSE
  ) +
  scale_fill_grey(end = 0.2, start = 0.8) +
  theme_bw()
```

The episode_days option sets the "gap" to look for, rather than the total permitted length of an episode. For example, LFT+ve on d-6, symptoms on d0, and PCR+ on d7 would be collected into a single episode.

## Episode variant

Next, we assign variants. There is no one-size-fits all approach that holds across pathogens, across studies, and across geographic regions. In this example we have PCR data available, and could parse as: if there is PCR positivity assign Delta, if there is no PCR testing assign Ancestral/Alpha. 

```{r}
cg_episodes <- cg_episodes %>%
  mutate(
    episode_variant =
      case_when(
        # "is an episode" & "PCR positive" -> Delta #
        (!is.na(episode_number)) & PCR == "Pos" ~ "Delta",
        # "is an episode" & "PCR unavailable" -> Anc/Delta #
        (!is.na(episode_number)) & PCR == "not tested" ~ "Anc/Alpha"
      )
  )
```

This gives a variant call on a single row of `cg`.

```{r}
cg_episodes %>%
  filter(!is.na(episode_variant)) %>%
  pull(episode_variant)
```

We can use `cg_annotate_episodes_fill()` to add this annotation to every row of the infection episode.

```{r}
cg_episodes <- cg_episodes %>%
  cg_annotate_episodes_fill(
    col_to_fill = episode_variant,
    col_to_return = episode_variant_filled,
    .direction = "updown",
    episode_numbers_col = episode_number
  )


cg_episodes %>%
  filter(!is.na(episode_variant_filled)) %>%
  pull(episode_variant_filled)
```

Variant calling rarely relies on a single assay. We advocate a pragmatic a tiered approach. If viral sequencing results are available we use those; if PCR only (Â± S gene target failure), we use those; and in the absence of molecular testing we compare episode start dates to prevailing variant trends. 

In principle:

- each assay result is reported in its own column in 1 (or more) experimental data tibbles

- build with `cg_assemble()`

- find episodes as above

- fill each of `episode_variant_by_seq`, `episode_variant_by_PCR` and `episode_variant_by_date`, and then fill for the whole episode, using `cg_annotate_episodes_fill()`.

- write a `case_when()` to combine over `episode_variant_by_seq_filled`, `episode_variant_by_PCR_filled` and `episode_variant_by_date_filled`. Which result trumps which other results is study-dependent.

... to return a column `episode_variant_summarised` as the combined, overall assignment (column naming strategy entirely up to the user).

# Summary

This vignette has provided examples of the cg_annotate family in action. If you are conducting a multi-pathogen study (RSV, flu, covid), then run a set of cg_annotate family functions for each pathogen - and you may wish to prefix the output columns eg `RSV_`,  `flu_` & `covid_`. As these have differing considerations for eg variants, chronogram leaves the cg_annotate family without an overall wrapper to let users easily omit unneeded annotations.

## SessionInfo

```{r}
sessionInfo()
```
