---
title: "Annotating infection episodes"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Annotating infection episodes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\VignetteDepends{stringr}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(dplyr)
library(stringr) # for str_detect()
library(chronogram)
```

This vignette describes how chronogram deals with episodes of infection.

Steps:

- assemble a chronogram (see `vignette("assembly")` for explainer)
- add infection data, as an `experiment data` tibble
- find and annotate these episodes

The vignette finishes by considering how to handle assigning variants to infection episodes.

## Build a chronogram

```{r}
## Fictional example data ##
data(smallstudy)

## 5 requirements ##
ids <- smallstudy$small_study_ids

start <- "01012020"
end <- "10102021"

meta <- smallstudy$small_study_metadata
ab <- smallstudy$small_study_Ab # here, we just have antibody data


## Make a chronogram ##
small_study <- chronogram_skeleton(
  ids = ids,
  start_date = start,
  end_date = end,
  ## change this to your ID column name ##
  col_ids = elig_study_id,
  ## change this to your date column name ##
  col_calendar_date = calendar_date
)

small_study <- chronogram(small_study, meta)

small_study <- cg_add_experiment(
  small_study,
  ab
)

small_study

## Use cg_add_experiment() for extra assays ##
```

## Some example infection data

```{r}
## Simulate some infection data ##
infections_to_add <- tibble::tribble(
  ~calendar_date, ~elig_study_id, ~LFT, ~PCR, ~symptoms,
  "01102020", "1", "pos", NA, NA,
  "11102020", "1", "pos", NA, "severe"
)
## Make calendar_date a date ##
infections_to_add$calendar_date <- lubridate::dmy(
  infections_to_add$calendar_date
)
## add to chronogram
small_study_inf <- cg_add_experiment(
  small_study,
  infections_to_add
)

small_study_inf

small_study_inf %>% arrange(PCR)
```

`small_study` is now a chrongram with infection episodes, and serological data.

## Infection episode parsing

The episode definition will depend on the study's objectives, design and hypotheses. For example, a study with weekly nasopharyngeal PCRs may wish to use only PCR tests to define an episode, so as to include asymptomatic infections. In this vignette, we assume the study's definition is to call any positive LFT or PCR or severe symptoms (even if not LFT/PCR +ve) to be counted as an episode. This is illustrative: your study will need its own solution. Chronogram provides generalised code, to allow the code implementation of an episode definition.

In the following code chunk, we're identifying episodes of infection baesd on specific strings within specified columns. In this dataset, we look for "pos" in column `LFT`, "Post" in column `PCR` and "^severe" in column `symptoms`. The presence of any one of those triggers an infection_episode being labelled. `cg_annotate_episodes_find()` searches backwards and forwards to scoop up different calendar dates into single episodes, as it is common for symptoms to precede test positivity (you can tweak this with `episode_days`).

```{r}
small_study_inf <- cg_annotate_episodes_find(small_study_inf,
  infection_cols = c("LFT", "PCR", "symptoms"),
  infection_present = c("pos", "Post", "^severe")
)

## summary statistics ##
summary(small_study_inf$episode_number)


## numbers of infection episodes per individual ##
small_study_inf %>%
  group_by(elig_study_id) %>%
  summarise(n = max(episode_number, na.rm = TRUE))

small_study_inf
```

You have the flexibility to configure `infection_cols` and `infection_present` to match any string on any column. The matching is performed as string matching. Care is required with matching exact text. Base R's `regex()` help has lots of examples of matching expressions including using wildcards. Below are a few examples of string matching problems.

```{r eval=F}
## Not run in vignette ... ##
## these are examples of text matching failures ##

## exact text matching ##
test2 <- cg_annotate_episodes_find(small_study_inf,
  infection_cols = c("LFT", "PCR", "symptoms"),
  infection_present = c("Pos", "Post", "^mild")
)
summary(test2$episode_number)

## empty strings will error (as they otherwise match everything) ##
test3a <- cg_annotate_episodes_find(small_study_inf,
  infection_cols = c("LFT", "PCR", "symptoms"),
  infection_present = c("pos", "Post", "")
)

## a 'random' string will not error ##
test3b <- cg_annotate_episodes_find(small_study_inf,
  infection_cols = c("LFT", "PCR", "symptoms"),
  infection_present = c("pos", "Post", "a")
)
```

## Variant calling

For SARS-CoV-2 studies, it is often desirable to report an infecting variant. The fundamental approach is: "A viral sequencing result trumps an inference based on S gene target failure (SGTF) which trumps an inference based on date of symptoms (or LFT) only." However, there are four ways this becomes more complex in the wild:

1. _Available data_

Not all studies will have symptom diaries, or SGTF data (perhaps only PCR positive/negative, or only used LFTs) and not all studies will sequence at least one swab from each infection episode. Chronogram should not mandate the presence of these data.

2. _Sequencing tools, and their results, develop over time_

As new variants evolve, emerge and are sequenced in enough numbers to be assigned their own lineage designations, PANGO and scorpio definitions need to be revised. As scorpio uses "constellations" of sentinel ("lineage defining") SNPs, the scorpio defintion can take a little longer to be released, PANGO is the bleeding-edge. Therefore XBB.1.5, for example, might be called `XBB.1.5` by PANGO and `Omicron-BA.2 like` by scorpio. Most studies would want the final assigned variant to be `XBB.1.5`. Conversely, many investigators would want to group all of `Delta` rather than have `AY.42` and `B.1.617.2`. This solvable in chronogram, but we have not hard-coded an approach (as there isn't a one-size-fits-all solution).

3. _The external inference data changes over time_

For example, assigning Alpha vs Delta based on date alone in the UK is pretty straightforward _today_, as there is a large amount of sequences available. At the time it was more challenging. In 2023, assigning variants based on date in the UK is very challenging - with over >10 Omicron sub-variants co-circulating. Chronogram does not assume it can do this for the end user.

4. _The returned data will map by calendar date_

In Chronogram, each data entry, such as a variant identification (`PANGO==XBB.1.5`), is timestamped by the calendar date when the swab was taken. An investigator will want `XBB.1.5` to be the variant call associated with the entire episode (consider: XBB.1.5+ve sequence result on d0, symptoms on d-5 and LFT+ve on d-4, this is an `XBB.1.5 episode spanning d-5 to d0`). Though the data might be scattered across different rows in Chronogram, a researcher would interpret this as a single XBB.1.5 episode that spanned from d-5 to d0.

Chronogram offers a flexible way to address point 4 i.e. tracking data by its calendar date. For point 1, this requires a unique coing. solution tailed to each study's dataset. For 2+3, study specific solutions are required, and for contemporaneous studies (eg does latest vaccine protect against current variant?), then a rolling review of the strategy is required.

### Variant dataset

Returning to our small example study, we now have SGTF data too. Dates of these episodes have been updated to be plausibly Ancestral infection with S gene presence, and Alpha infection with SGTF.

First, build chronogram as previously.

```{r}
## Simulate some variant data ##
variants_to_add <- tibble::tribble(
  ~calendar_date, ~elig_study_id, ~LFT, ~PCR, ~SGTF, ~symptoms,
  "01062020", "1", "pos", NA, "S present", NA,
  "01022021", "1", NA, NA, NA, "severe",
  "11022021", "1", "pos", NA, "S fail", "severe"
)

## Make calendar_date a date ##
variants_to_add$calendar_date <- lubridate::dmy(
  variants_to_add$calendar_date
)
## add to chronogram ##
small_study.variants <- cg_add_experiment(small_study, variants_to_add)

small_study.variants <- cg_annotate_episodes_find(small_study.variants,
  infection_cols = c("LFT", "PCR", "symptoms"),
  infection_present = c("pos", "Post", "^severe")
)
```

Next, we define a new column `episode_filled_SGTF` which is the first non-NA value for SGTF within each episode (solving issue 4 above).

### Fill episode data

Within each episode, "spread" an experimental result to every row. In `tidyverse`, this is performed with `tidyr::fill()`, and so we have adopted a similar name for the chronogram function. `Tidyr::fill()` has an argument `.direction` which allows filling upwards or downwards (or one then the other). This is useful where there are several results: eg. S gene may be present at the first swab in an episode, and once infection cleared and PCR negative, a second swab could show SGTF (it would also be negative for the other PCR products). Using the earliest, S gene present result is useful. However, in other instances, for example the burden of viral polymorphisms, that may peak at the last measurement in that episode. `.direction` is passed onwards to the wrapped `tidyr::fill()`.

```{r}
small_study.variants <- cg_annotate_episodes_fill(small_study.variants,
  col_to_fill = SGTF,
  col_to_return = episode_filled_SGTF,
  .direction = "downup",
  episode_numbers_col = episode_number
)

small_study.variants$episode_filled_SGTF
```

Each variant identifying assay needs to be `filled` as above.


### Assign episode data

Now we can use a `case_when()` within each group of id + episode_number.

```{r}
small_study.variants <- small_study.variants %>%
  mutate(episode_variant = case_when(
    str_detect(episode_filled_SGTF, "S present") ~ "Ancestral",
    str_detect(episode_filled_SGTF, "S fail") ~ "Alpha"
  ))


## How many days are each episode? ##
## Alpha = 11 days long; Ancestral = 1 day long ##
small_study.variants %>%
  group_by(elig_study_id, episode_number, episode_variant) %>%
  tally()
```

This case_when() is straightforward.

How would you use PANGO before SGTF?

>
>__HINT__
> `case_when(!is.na(episode_filled_PANGO) ~ episode_filled_PANGO, ...)`
>

How would you adjust with emergence of Delta, in a study without viral sequencing? (S present).

>
>__HINT__
> `case_when(str_detect(episode_filled_SGTF, "S present") & calendar_date > lubridate::dmy("01092021") ~ "Delta", ...)`
>

## Summary

We have added infection data to chronogram, and summarised variant calls, with flexible and scalable logic. 

## SessionInfo

```{r}
sessionInfo()
```

