<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chronogram annotation • chronogram</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Chronogram annotation">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">chronogram</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/chronogram.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/annotate_episodes.html">Annotating infection episodes</a></li>
    <li><a class="dropdown-item" href="../articles/annotation.html">Chronogram annotation</a></li>
    <li><a class="dropdown-item" href="../articles/assembly.html">Chronogram assembly</a></li>
    <li><a class="dropdown-item" href="../articles/SQL_assembly.html">Chronogram assembly from SQL</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Chronogram annotation</h1>
            
      

      <div class="d-none name"><code>annotation.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>The chronogram package provides a family of functions to annotate a
chronogram. These all start <code>cg_annotate_</code>. This vignette
explains how to use these annotation functions. Before using this
vignette, consult the <code><a href="../articles/assembly.html">vignette("assembly")</a></code>.</p>
</div>
<div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://franciscrickinstitute.github.io/chronogram/">chronogram</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre></div>
<p>We will use the example pre-built chronogram, introduced in the
<code><a href="../articles/assembly.html">vignette("assembly")</a></code>, and add on some example infection
data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">built_smallstudy</span><span class="op">)</span></span>
<span><span class="va">cg</span> <span class="op">&lt;-</span> <span class="va">built_smallstudy</span><span class="op">$</span><span class="va">chronogram</span></span>
<span></span>
<span><span class="co">## Simulate some infection data ##</span></span>
<span><span class="va">infections_to_add</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span><span class="va">calendar_date</span>, <span class="op">~</span><span class="va">elig_study_id</span>, <span class="op">~</span><span class="va">LFT</span>, <span class="op">~</span><span class="va">PCR</span>, <span class="op">~</span><span class="va">symptoms</span>,</span>
<span>  <span class="st">"01102020"</span>, <span class="st">"1"</span>, <span class="st">"pos"</span>, <span class="st">"not tested"</span>, <span class="cn">NA</span>,</span>
<span>  <span class="st">"11102020"</span>, <span class="st">"1"</span>, <span class="st">"pos"</span>, <span class="cn">NA</span>, <span class="st">"severe"</span>,</span>
<span>  <span class="st">"11042021"</span>, <span class="st">"2"</span>, <span class="st">"pos"</span>, <span class="st">"not tested"</span>, <span class="st">"mild"</span>,</span>
<span>  <span class="st">"11082021"</span>, <span class="st">"2"</span>, <span class="st">"pos"</span>, <span class="st">"Pos"</span>, <span class="st">"mild"</span>,</span>
<span>  <span class="st">"11112021"</span>, <span class="st">"2"</span>, <span class="st">"pos"</span>, <span class="st">"Pos"</span>, <span class="cn">NA</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Make calendar_date a date ##</span></span>
<span><span class="va">infections_to_add</span><span class="op">$</span><span class="va">calendar_date</span> <span class="op">&lt;-</span> <span class="fu">lubridate</span><span class="fu">::</span><span class="fu"><a href="https://lubridate.tidyverse.org/reference/ymd.html" class="external-link">dmy</a></span><span class="op">(</span></span>
<span>  <span class="va">infections_to_add</span><span class="op">$</span><span class="va">calendar_date</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## add to chronogram</span></span>
<span><span class="va">cg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cg_add_experiment.html">cg_add_experiment</a></span><span class="op">(</span></span>
<span>  <span class="va">cg</span>,</span>
<span>  <span class="va">infections_to_add</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="outline">Outline<a class="anchor" aria-label="anchor" href="#outline"></a>
</h3>
<p>Annotation is required to allow the selection sub-cohorts of
individuals (and corresponding dates) that are relevant to test your
biological hypothesis.</p>
<div class="section level4">
<h4 id="the-annotation-process">The annotation process<a class="anchor" aria-label="anchor" href="#the-annotation-process"></a>
</h4>
<div class="section level5">
<h5 id="for-vaccines-use-cg_annotate_vaccines_count">1. For vaccines, use <code>cg_annotate_vaccines_count()</code><a class="anchor" aria-label="anchor" href="#for-vaccines-use-cg_annotate_vaccines_count"></a>
</h5>
<p>– adds a column which counts the number of vaccines each participant
has received over time.</p>
<p>– includes a “star” system, to allow the first few days after a
vaccine to be easily identified. For example, 24hrs after the first dose
of a vaccine (“1star”) is probably not biologically reflect of that
dose’s effect. The user can set the value of days to “star” to suit
their analysis.</p>
</div>
<div class="section level5">
<h5 id="for-episodes-use-the-cg_annotate_episodes_-family">2. For episodes, use the <code>cg_annotate_episodes_</code>
family<a class="anchor" aria-label="anchor" href="#for-episodes-use-the-cg_annotate_episodes_-family"></a>
</h5>
<p>These perform sequential steps:</p>
<ol style="list-style-type: lower-roman">
<li>
<code><a href="../reference/cg_annotate_episodes_find.html">cg_annotate_episodes_find()</a></code> is used to identify
episodes of infection. This <em>reads</em> across several different
experiment data types, that are specified in the function call: eg.
symptoms, PCRs, LFTs for SARS-CoV-2. Additionally,
<code><a href="../reference/cg_annotate_episodes_find.html">cg_annotate_episodes_find()</a></code> looks backwards and forwards in
time to find episodes, as commonly the molecular testing and symptoms do
not commence on the same day.</li>
</ol>
<p>In some studies, seroconversion is the method of finding episodes (eg
seroconversion on anti-N IgG in SARS-CoV-2). Seroconversion episodes can
be found with <code><a href="../reference/cg_annotate_episodes_find_seroconversion.html">cg_annotate_episodes_find_seroconversion()</a></code>.
As these provide different (often much, much larger) uncertainty around
the episode start, chronogram handles this in a separate function.</p>
<ol start="2" style="list-style-type: lower-roman">
<li><p><code><a href="../reference/cg_annotate_episodes_fill.html">cg_annotate_episodes_fill()</a></code> allow filling of
specific columns for each episode (details below).</p></li>
<li><p>count the number of episodes (cf
<code><a href="../reference/cg_annotate_vaccines_count.html">cg_annotate_vaccines_count()</a></code>)</p></li>
</ol>
</div>
<div class="section level5">
<h5 id="for-exposures-use-cg_exposure_count-and-cg_antigenic_history">3. For exposures, use <code>cg_exposure_count()</code> and
<code>cg_antigenic_history()</code><a class="anchor" aria-label="anchor" href="#for-exposures-use-cg_exposure_count-and-cg_antigenic_history"></a>
</h5>
<p>Exposures can be either from vaccine doses or from infection
episodes. This function provides a cumulative counter for each
individual’s personal history.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="annotating-vaccines">1. Annotating vaccines<a class="anchor" aria-label="anchor" href="#annotating-vaccines"></a>
</h3>
<p><code><a href="../reference/cg_annotate_vaccines_count.html">cg_annotate_vaccines_count()</a></code> requires that metadata
columns follow this pattern:</p>
<ul>
<li><p>dose_1, dose_2, dose_3, …, dose_i</p></li>
<li><p>date_dose_1, date_dose_2, date_dose_3, …, date_dose_i</p></li>
</ul>
<p>The {dose} and {date_dose} prefixes should be provided to the
function, as shown in the chunk below. You might envisage a study with
{sarscov2_dose}, {sarscov2_date_dose}, {influenza_dose} &amp;
{influenza_date_dose} for which two runs of
<code><a href="../reference/cg_annotate_vaccines_count.html">cg_annotate_vaccines_count()</a></code> would be needed.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cg1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cg_annotate_vaccines_count.html">cg_annotate_vaccines_count</a></span><span class="op">(</span></span>
<span>  <span class="va">cg</span>,</span>
<span>  <span class="co">## the prefix to the dose columns: ##</span></span>
<span>  dose <span class="op">=</span> <span class="va">dose</span>,</span>
<span>  <span class="co">## the output column name: ##</span></span>
<span>  dose_counter <span class="op">=</span> <span class="va">dose_number</span>,</span>
<span>  <span class="co">## the prefix to the date columns: ##</span></span>
<span>  vaccine_date_stem <span class="op">=</span> <span class="va">date_dose</span>,</span>
<span>  <span class="co">## use 14d to 'star' after a dose ##</span></span>
<span>  intermediate_days <span class="op">=</span> <span class="fl">14</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Using stem: date_dose</span></span>
<span><span class="co">#&gt; Found vaccine dates</span></span>
<span><span class="co">#&gt; date_dose_1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; date_dose_2</span></span>
<span></span>
<span><span class="co">## plot over time ##</span></span>
<span><span class="va">cg1</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>      x <span class="op">=</span> <span class="va">calendar_date</span>,</span>
<span>      y <span class="op">=</span> <span class="va">elig_study_id</span>,</span>
<span>      fill <span class="op">=</span> <span class="va">dose_number</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span>height <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_grey.html" class="external-link">scale_fill_grey</a></span><span class="op">(</span>end <span class="op">=</span> <span class="fl">0.2</span>, start <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="annotation_files/figure-html/unnamed-chunk-3-1.png" width="576"></p>
<p>In the above plot, the character vector <code>dose_number</code> is
coerced to a factor. The resulting levels of the factor are
counter-intuitive. To improve the plot, you can manually specify the
factor:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cg1</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>dose_number <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">dose_number</span>,</span>
<span>    levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"0"</span>,</span>
<span>      <span class="st">"1star"</span>,</span>
<span>      <span class="st">"1"</span>,</span>
<span>      <span class="st">"2star"</span>,</span>
<span>      <span class="st">"2"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>      x <span class="op">=</span> <span class="va">calendar_date</span>,</span>
<span>      y <span class="op">=</span> <span class="va">elig_study_id</span>,</span>
<span>      fill <span class="op">=</span> <span class="va">dose_number</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span>height <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_grey.html" class="external-link">scale_fill_grey</a></span><span class="op">(</span>end <span class="op">=</span> <span class="fl">0.2</span>, start <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="annotation_files/figure-html/unnamed-chunk-4-1.png" width="576"></p>
<p>The <code>dose_number</code> column is retained as character vector
rather than converting to a factor. There is the possibility for
mishandling if a factor: <code>as.numeric(dose_number) == 2</code>,
refers to the situation <code>dose == 1</code>.</p>
</div>
<div class="section level3">
<h3 id="annotating-episodes">2. Annotating episodes<a class="anchor" aria-label="anchor" href="#annotating-episodes"></a>
</h3>
<p>The first step is to find episodes using
<code><a href="../reference/cg_annotate_episodes_find.html">cg_annotate_episodes_find()</a></code></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cg2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cg_annotate_episodes_find.html">cg_annotate_episodes_find</a></span><span class="op">(</span></span>
<span>  <span class="va">cg1</span>,</span>
<span>  infection_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LFT"</span>, <span class="st">"PCR"</span>, <span class="st">"symptoms"</span><span class="op">)</span>,</span>
<span>  infection_present <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pos"</span>, <span class="st">"Post"</span>, <span class="st">"^severe"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Parsed: infection_cols and infection_present</span></span>
<span><span class="co">#&gt;           </span></span>
<span><span class="co">#&gt; Searching in the [[column]], for the "text":</span></span>
<span><span class="co">#&gt; stringr::str_detect(.data[["LFT"]], "pos") ~ "yes"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; stringr::str_detect(.data[["PCR"]], "Post") ~ "yes"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; stringr::str_detect(.data[["symptoms"]], "^severe") ~ "yes"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ...detecting will be exact.</span></span>
<span><span class="co">#&gt;           Capitals, spelling etc must be precise</span></span>
<span><span class="co">#&gt; calendar_date elig_study_id episode_number</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(calendar_date, elig_study_id)`</span></span>
<span></span>
<span><span class="va">cg2</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>dose_number <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">dose_number</span>,</span>
<span>    levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"1star"</span>, <span class="st">"1"</span>, <span class="st">"2star"</span>, <span class="st">"2"</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>      x <span class="op">=</span> <span class="va">calendar_date</span>,</span>
<span>      y <span class="op">=</span> <span class="va">elig_study_id</span>,</span>
<span>      fill <span class="op">=</span> <span class="va">dose_number</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span>height <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co">## now add text to show when episodes occur ##</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">episode_number</span><span class="op">)</span>,</span>
<span>    data <span class="op">=</span> <span class="va">.</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">elig_study_id</span>, <span class="va">episode_number</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"red"</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_grey.html" class="external-link">scale_fill_grey</a></span><span class="op">(</span>end <span class="op">=</span> <span class="fl">0.2</span>, start <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 3 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_text()`).</span></span></code></pre></div>
<p><img src="annotation_files/figure-html/unnamed-chunk-5-1.png" width="576"></p>
<p>Note that participant 1 has a single infection episode in November
2020:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">infections_to_add</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">elig_study_id</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">calendar_date</th>
<th align="left">elig_study_id</th>
<th align="left">LFT</th>
<th align="left">PCR</th>
<th align="left">symptoms</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">2020-10-01</td>
<td align="left">1</td>
<td align="left">pos</td>
<td align="left">not tested</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">2020-10-11</td>
<td align="left">1</td>
<td align="left">pos</td>
<td align="left">NA</td>
<td align="left">severe</td>
</tr>
</tbody>
</table>
<p>We could use a shorter time frame to look forwards and backwards
in:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cg3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cg_annotate_episodes_find.html">cg_annotate_episodes_find</a></span><span class="op">(</span></span>
<span>  <span class="va">cg1</span>,</span>
<span>  infection_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LFT"</span>, <span class="st">"PCR"</span>, <span class="st">"symptoms"</span><span class="op">)</span>,</span>
<span>  infection_present <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pos"</span>, <span class="st">"Post"</span>, <span class="st">"^severe"</span><span class="op">)</span>,</span>
<span>  <span class="co">## change episode days interval to 7 days ##</span></span>
<span>  episode_days <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Parsed: infection_cols and infection_present</span></span>
<span><span class="co">#&gt;           </span></span>
<span><span class="co">#&gt; Searching in the [[column]], for the "text":</span></span>
<span><span class="co">#&gt; stringr::str_detect(.data[["LFT"]], "pos") ~ "yes"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; stringr::str_detect(.data[["PCR"]], "Post") ~ "yes"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; stringr::str_detect(.data[["symptoms"]], "^severe") ~ "yes"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ...detecting will be exact.</span></span>
<span><span class="co">#&gt;           Capitals, spelling etc must be precise</span></span>
<span><span class="co">#&gt; calendar_date elig_study_id episode_number</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(calendar_date, elig_study_id)`</span></span>
<span></span>
<span><span class="va">cg3</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>dose_number <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">dose_number</span>,</span>
<span>    levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"1star"</span>, <span class="st">"1"</span>, <span class="st">"2star"</span>, <span class="st">"2"</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>      x <span class="op">=</span> <span class="va">calendar_date</span>,</span>
<span>      y <span class="op">=</span> <span class="va">elig_study_id</span>,</span>
<span>      fill <span class="op">=</span> <span class="va">dose_number</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span>height <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co">## now add text to show when episodes occur ##</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">episode_number</span><span class="op">)</span>,</span>
<span>    data <span class="op">=</span> <span class="va">.</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">elig_study_id</span>, <span class="va">episode_number</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"red"</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_grey.html" class="external-link">scale_fill_grey</a></span><span class="op">(</span>end <span class="op">=</span> <span class="fl">0.2</span>, start <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 3 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_text()`).</span></span></code></pre></div>
<p><img src="annotation_files/figure-html/unnamed-chunk-7-1.png" width="576"></p>
<p>The episode_days option sets the “gap” to look for, rather than the
total permitted length of an episode. For example, LFT+ve on d-6,
symptoms on d0, and PCR+ on d7 would be collected into a single
episode.</p>
<p>Next, we assign variants. There is no one-size-fits all approach that
holds across pathogens, across studies, and across geographic regions.
In this example we have PCR data available, and could parse as: if there
is PCR positivity assign Delta, if there is no PCR testing assign
Ancestral/Alpha.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cg4</span> <span class="op">&lt;-</span> <span class="va">cg2</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    episode_variant <span class="op">=</span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span></span>
<span>        <span class="co"># "is an episode" &amp; "PCR positive" -&gt; Delta #</span></span>
<span>        <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">episode_number</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">PCR</span> <span class="op">==</span> <span class="st">"Pos"</span> <span class="op">~</span> <span class="st">"Delta"</span>,</span>
<span>        <span class="co"># "is an episode" &amp; "PCR unavailable" -&gt; Anc/Delta #</span></span>
<span>        <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">episode_number</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">PCR</span> <span class="op">==</span> <span class="st">"not tested"</span> <span class="op">~</span> <span class="st">"Anc/Alpha"</span></span>
<span>      <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>This gives a variant call on a single row of <code>cg</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cg4</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">episode_variant</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">episode_variant</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Anc/Alpha" "Anc/Alpha" "Delta"</span></span></code></pre></div>
<p>We can use <code><a href="../reference/cg_annotate_episodes_fill.html">cg_annotate_episodes_fill()</a></code> to add this
annotation to every row of the infection episode.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cg4</span> <span class="op">&lt;-</span> <span class="va">cg4</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/cg_annotate_episodes_fill.html">cg_annotate_episodes_fill</a></span><span class="op">(</span></span>
<span>    col_to_fill <span class="op">=</span> <span class="va">episode_variant</span>,</span>
<span>    col_to_return <span class="op">=</span> <span class="va">episode_variant_filled</span>,</span>
<span>    .direction <span class="op">=</span> <span class="st">"updown"</span>,</span>
<span>    episode_numbers_col <span class="op">=</span> <span class="va">episode_number</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(elig_study_id, episode_number)`</span></span>
<span></span>
<span></span>
<span><span class="va">cg4</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">episode_variant_filled</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">episode_variant_filled</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Anc/Alpha" "Anc/Alpha" "Anc/Alpha" "Anc/Alpha" "Anc/Alpha" "Anc/Alpha"</span></span>
<span><span class="co">#&gt;  [7] "Anc/Alpha" "Anc/Alpha" "Anc/Alpha" "Anc/Alpha" "Anc/Alpha" "Anc/Alpha"</span></span>
<span><span class="co">#&gt; [13] "Delta"</span></span></code></pre></div>
<p>Variant calling rarely relies on a single assay. We advocate a
pragmatic a tiered approach. If viral sequencing results are available
we use those; if PCR only (± S gene target failure), we use those; and
in the absence of molecular testing we compare episode start dates to
prevailing variant trends. Give each their own
<code>episode_variant_by_seq</code>, <code>episode_variant_by_PCR</code>
and <code>episode_variant_by_date</code>, and then fill
<code>*_filled</code> for the whole episode. And then apply a further
<code><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when()</a></code> to combine over
<code>episode_variant_by_seq_filled</code>,
<code>episode_variant_by_PCR_filled</code> and
<code>episode_variant_by_date_filled</code>. We adopt
<code>episode_variant_summarised</code> as the combined, overall
assignment, but the user is free to change.</p>
<p>Sequencing interpretation is its own sub-discipline, which we have
simplified in this vignette. For SARS-CoV-2, there are two main methods
of assigning a variant: PANGO and scorpio, and these, on occasion,
report different lineages particularly for newly emerging viruses. We
keep both: <code>episode_variant_by_pango</code> and
<code>episode_variant_by_scorpio</code>.</p>
<p>Finally, we can count the number of episodes that each individual has
experienced.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cg4</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/cg_annotate_episodes_count.html">cg_annotate_episodes_count</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">count_col</span><span class="op">)</span></span>
<span><span class="co">#&gt;    [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;   [38] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;   [75] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [112] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [149] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [186] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [223] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [260] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [297] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [334] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [371] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [408] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [445] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [482] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [519] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [556] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [593] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</span></span>
<span><span class="co">#&gt;  [630] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt;  [667] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt;  [704] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt;  [741] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt;  [778] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt;  [815] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt;  [852] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt;  [889] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt;  [926] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt;  [963] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt; [1000] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt; [1037] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt; [1074] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt; [1111] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt; [1148] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt; [1185] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt; [1222] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt; [1259] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></span>
<span><span class="co">#&gt; [1296] 2 2 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1333] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1370] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1407] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1444] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1481] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1518] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1555] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1592] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1629] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1666] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1703] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1740] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1777] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1814] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1851] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1888] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span><span class="co">#&gt; [1925] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="annotating-exposures">3. Annotating exposures<a class="anchor" aria-label="anchor" href="#annotating-exposures"></a>
</h3>
<p>Annotation of exposures is the cumulative sum of infection +
vaccination histories. These annotations are provided by
<code>cg_exposure_count()</code> and
<code>cg_antigenic_history()</code>.</p>
<p><code>cg_exposure_count()</code> takes the column names for episode
number, dose number and seroconversion episode numbers, to calculate a
per day, per individual running count of exposures.</p>
<p><code><a href="../reference/cg_annotate_antigenic_history.html">cg_annotate_antigenic_history()</a></code> provides a character
vector of length 1 to summarise each person’s course over the study.</p>
<p>These are both best explored with an example, and a plot.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cg_exposures</span> <span class="op">&lt;-</span> <span class="va">cg4</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/cg_annotate_exposures_count.html">cg_annotate_exposures_count</a></span><span class="op">(</span></span>
<span>  episode_number <span class="op">=</span> <span class="va">episode_number</span>,</span>
<span>  dose_number <span class="op">=</span> <span class="va">dose_number</span>,</span>
<span>  <span class="co">## we have not considered episodes of seroconversion</span></span>
<span>  N_seroconversion_episode_number <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">cg_exposures</span> <span class="op">&lt;-</span> <span class="va">cg_exposures</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    episode_variant_summarised <span class="op">=</span></span>
<span>      <span class="va">episode_variant_filled</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/cg_annotate_antigenic_history.html">cg_annotate_antigenic_history</a></span><span class="op">(</span></span>
<span>    episode_number <span class="op">=</span> <span class="va">episode_number</span>,</span>
<span>    dose_number <span class="op">=</span> <span class="va">dose_number</span>,</span>
<span>    episode_variant_summarised <span class="op">=</span> <span class="va">episode_variant_summarised</span>,</span>
<span>    ag_col <span class="op">=</span> <span class="va">antigenic_history</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot ##</span></span>
<span><span class="va">top_panel</span> <span class="op">&lt;-</span> <span class="va">cg_exposures</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">calendar_date</span>, </span>
<span>         <span class="va">exposure_number</span>,</span>
<span>         <span class="va">elig_study_id</span>,</span>
<span>         <span class="va">antigenic_history</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">calendar_date</span>, y <span class="op">=</span> <span class="va">exposure_number</span>,</span>
<span>    col <span class="op">=</span> <span class="va">elig_study_id</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">antigenic_history</span> <span class="op">~</span> <span class="va">.</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">swimmers_panel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cg_plot_meta.html">cg_plot_meta</a></span><span class="op">(</span><span class="va">cg_exposures</span>,</span>
<span>  visit <span class="op">=</span> <span class="va">serum_Ab_S</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co">## set the axes to match top_panel ##</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">cg_exposures</span><span class="op">$</span><span class="va">calendar_date</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">cg_exposures</span><span class="op">$</span><span class="va">calendar_date</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_y_discrete</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Function provided to illustrate chronogram -&gt;</span></span>
<span><span class="co">#&gt;           ggplot2 interface.</span></span>
<span><span class="co">#&gt; Function assumes the</span></span>
<span><span class="co">#&gt;           presence of {dose_1, date_dose_1, dose_2, date_dose_2}</span></span>
<span><span class="co">#&gt;           columns.</span></span>
<span><span class="co">#&gt;           Users are likely to want to write their own,</span></span>
<span><span class="co">#&gt;           study-specific applications</span></span>
<span></span>
<span><span class="va">top_panel</span> <span class="op">/</span> <span class="va">swimmers_panel</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>    strip.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>    strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="annotation_files/figure-html/unnamed-chunk-12-1.png" width="576"></p>
<p>The plot above shows how each infection, or vaccination impacts the
exposure number. For participant 3, their only exposures are
vaccination, whereas both participants 1 and 2 have additional exposures
from infection. By the end of this example, participant 2 has
experienced 4 encounters with Spike, participant 1 has had 3 encounters
and participant 3 just two from vaccination alone.</p>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>This vignette has provided examples of the cg_annotate family in
action. If you are conducting a multi-pathogen study (RSV, flu, covid),
then run a set of cg_annotate family functions for each pathogen - and
you may wish to prefix the output columns eg <code>RSV_</code>,
<code>flu_</code> &amp; <code>covid_</code>. As these have differing
considerations for eg variants, chronogram leaves the cg_annotate family
without an overall wrapper to let users easily omit unneeded
annotations.</p>
<div class="section level3">
<h3 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h3>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] patchwork_1.2.0  ggplot2_3.5.1    dplyr_1.1.4      chronogram_1.0.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] gtable_0.3.5      jsonlite_1.8.8    highr_0.11        compiler_4.4.1   </span></span>
<span><span class="co">#&gt;  [5] tidyselect_1.2.1  stringr_1.5.1     tidyr_1.3.1       jquerylib_0.1.4  </span></span>
<span><span class="co">#&gt;  [9] systemfonts_1.1.0 scales_1.3.0      textshaping_0.4.0 yaml_2.3.9       </span></span>
<span><span class="co">#&gt; [13] fastmap_1.2.0     R6_2.5.1          labeling_0.4.3    generics_0.1.3   </span></span>
<span><span class="co">#&gt; [17] knitr_1.48        tibble_3.2.1      desc_1.4.3        munsell_0.5.1    </span></span>
<span><span class="co">#&gt; [21] lubridate_1.9.3   bslib_0.7.0       pillar_1.9.0      rlang_1.1.4      </span></span>
<span><span class="co">#&gt; [25] utf8_1.2.4        stringi_1.8.4     cachem_1.1.0      xfun_0.45        </span></span>
<span><span class="co">#&gt; [29] fs_1.6.4          sass_0.4.9        timechange_0.3.0  cli_3.6.3        </span></span>
<span><span class="co">#&gt; [33] pkgdown_2.1.0     withr_3.0.0       magrittr_2.0.3    digest_0.6.36    </span></span>
<span><span class="co">#&gt; [37] grid_4.4.1        lifecycle_1.0.4   vctrs_0.6.5       evaluate_0.24.0  </span></span>
<span><span class="co">#&gt; [41] glue_1.7.0        farver_2.1.2      ragg_1.3.2        fansi_1.0.6      </span></span>
<span><span class="co">#&gt; [45] colorspace_2.1-0  purrr_1.0.2       rmarkdown_2.27    tools_4.4.1      </span></span>
<span><span class="co">#&gt; [49] pkgconfig_2.0.3   htmltools_0.5.8.1</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Edward Carr, David Greenwood, Marianne Shawe-Taylor.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
